mlir_graphblas/mlir_builder.py

```diff
@@ -7,54 +7,125 @@
+    def num_values(self) -> int:
+        return len(self.var_types)
+
     @property
-    def var_string(self) -> str:
-        return f"%{self.var_name}"
+    def var_type(self) -> str:
+        if not len(self.var_types) == 1:
+            raise TypeError(f"{self} has multiple types.")
+        var_type, = self.var_types
+        return var_type
+
+    def __repr__(self) -> str:
+        attributes_string = ', '.join(f'{k}={repr(self.__dict__[k])}' for k in sorted(self.__dict__.keys()))
+        return f'{self.__class__.__name__}({attributes_string})'
+    
+    def assign_string(self) -> str:
+        # TODO specify number of expected returned values as arg and sanity check
+        answer = f"%{self.var_name}"
+        if self.num_values != 1:
+            answer += f":{self.num_values}"
+        return answer
+
+    def access_string(self, index: Optional[int] = None) -> str:
+        answer = f"%{self.var_name}"
+        if index is None:
+            if self.num_values != 1:
+                raise RuntimeError("Attempting to use variable containing multiple values as a singleton variable.")
+        elif not isinstance(index, int):
+            raise LookupError(f"{self.__class__} only supports integer indexing.")
+        elif index<0 or self.num_values<=index:
+            raise IndexError(f"Index must be an integer in the range [0, {self.num_values}).")
+        else:
+            answer += f"#{index}"
+        return answer
+
+    def __getitem__(self, index: int) -> Tuple["MLIRVar", int]:
+        # Convenience function for use with methods like MLIRFunctionBuilder.return_vars
+        if not isinstance(index, int):
+            raise TypeError(f"{self.__class__.__name__} indices must be integers, not {type(index)}.")
+        return (self, index)
 
 
 class MLIRFunctionBuilder(BaseFunction):
 
-    # TODO consider making this class have a method to return a BaseFunction and not be a subclass of BaseFunction
-    # In other words, make it a "builder" that creates BaseFunction subclasses or instances instead of being a BaseFunction itself
+    # TODO consider making this class have a method to return a BaseFunction
+    # and not be a subclass of BaseFunction
+    # In other words, make it a "builder" that creates BaseFunction subclasses
+    # or instances instead of being a BaseFunction itself
+    # TODO Consider defining all the BaseFunction instances in functions.py via
+    # this class.
 
+    # TODO move this indentation to some parent class
+    default_indentation_size = 6
+    indentation_delta_size = 2
     function_wrapper_text = jinja2.Template(
-        """
-      func {% if private_func %}private{% endif %}@{{func_name}}({{signature}}) -> {{return_type}} {
-        {{statements}}
-      }
-"""
-    )  # TODO make this hard-coded indentation not hard-coded
+        "\n"+
+        (" "*default_indentation_size)+"func {% if private_func %}private{% endif %}@{{func_name}}({{signature}}) -> {{return_type}} {"+
```

Comment:
It might be a good idea to rename `return_type` to `return_types` since there can be multiple return types. I'll rename this.